schema-layer/data-structures/hashmap.md

```diff
@@ -0,0 +1,292 @@
+
+Each node in an IPLD HashMap is stored in a distinct IPLD block and CIDs are used for child node links.
+
+## Structure
+
+### Parameters
+
+Configurable parameters for any given IPLD HashMap:
+
+* `hashAlg`: The hash algorithm applied to keys in order to evenly distribute entries throughout the data structure. The algorithm is chosen based on speed, byte-size and randomness properties (but it must be available to the reader, hence the need for shared defaults, see below).
+* `bitWidth`: The number of bits to use at each level of the data structure for determining the index of the entry or a link to the next level of the data structure to continue searching. The equation `2`<sup>`bitWidth`</sup> yields the arity of the HashMap nodes, i.e. the number of storage locations for buckets and/or links to child nodes.
+* `bucketSize`: The maximum array size of entry storage buckets such that exceeding `bucketSize` causes the creation of a new child node to replace entry storage.
+
+### Node properties
+
+Each node in a HashMap data structure contains:
+
+* `data`: An array, with a length of one to `2`<sup>`bitWidth`</sup>.
+* `map`: A bitfield where the first `2`<sup>`bitWidth`</sup> bits are used to indicate whether a bucket or child node link is present at each possible index of the node.
+
+An important property of a HAMT is that the `data` array only contains active elements. Indexes in a node that do not contain any values (in buckets or links to child nodes) are not stored and the `map` bitfield is used to determine the `data` whether values are present and the array index of present values using a [`popcount()`](https://en.wikipedia.org/wiki/Hamming_weight). This allows us to store a maximally compacted `data` array for each node.
+
+### Schema
+
+The **root block** of an IPLD HashMap contains the same properties as all other blocks, in addition to configuration data that dictates how the algorithm below traverses and mutates the data structure.
+
+See [IPLD Schemas](/schema-layer/schemas) for a definition of this format.
+
+```ipldsch
+# Root node layout
+type HashMapRoot struct {
+  hashAlg String
+  bitWidth Int
+  bucketSize Int
+  map Bytes
+  data [ Element ]
+}
+
+# Non-root node layout
+type HashMapNode struct {
+  map Bytes
+  data [ Element ]
+}
+
+type Element union {
+  | Link "link"
```

Comment:
* We may want to call this a "shard" (this is a keyed union).
* I'm not sure if we should _require_ that this be a link. Like the UnixFS-2 proposal, it would be nice if we could abstract over block boundaries" (but this SHOULD be a link).
* If we abstract over block boundaries, I wonder if we even need buckets? We can literally say: If there are fewer than 3 values in a subdag, erase block boundaries. To weird?

Reply:
I'm not a fan of the term "shard" but am not strongly tied to "link" either.

@mikeal you've talked a lot about inlining entire UnixFSv2 structures into a single block, does that requirement extend to the case where you're needing a HAMT to store directories, or is there some upper size bound where that's not allowed? If you need this data structure to be inlinable then yes, we'll need to do something about this "link" piece.

In terms of having a heuristic for automatic inlining of child nodes, (a) that would make the case for buckets harder to sustain so they'd probably need to go and (b) we need to figure out how to retain canonical form. Is it as easy as applying the same rules as applied here to buckets? "if child node has <= `X`, inline it"?

There's a broader issue here of how to pack tree structures into blocks that I keep bumping into with collections, particularly sorted collections, but that might be a discussion for elsewhere.

Reply:
## For Reads

If you avoid defining where links are and build the read side entirely path based you end up with the flexibility to inline wherever you want. With IPLD Generics this concept ends up extending to all of the types, so even something like binary data could be a single inlined binary object or a big multi-block binary object. 

## For Writes

The heuristics here are a bit tricky. You don’t have *great* knowledge about the eventual size of the block before you encode it in order to decide when to inline and when to not inline. 

For files in the experimental `unixfsv2` implementation we use the `size` of the underlying binary data, and target about half the max-block size we’re looking for, and this `size` property accumulates and bubbles up through the directories as well (but we don’t currently shard directories in a HAMT). The obvious problem with this method is that if you have a lot of small files with large file names the encoding overhead is much larger than the size.

But, since the read size is completely agnostic of block boundaries we have room to improve all of this logic in the future.

Reply:
OK so for the purpose of this spec we still have to define `Link` in the schema even if it's a union of some kind (I know you don't like that @mikeal but we need to be able to say clearly where these kinds are in a block and a Link is an important one) but we should allow for fully inline data structures too.

Currently we have:

```ipldsch
type Element union {
  | Link "link"
  | Bucket "bucket"
} representation keyed
```

I'm going to have to involve @warpfork on this one because we lack the ability to define that one named property could be either a kind (Link) or a complex map representation of some other type (HashMapNode), so I can't just rename "link" to "child" and stick either a Link or HashMapNode in there, I have to do something like this:

```ipldsch
type Element union {
  | Child "child"
  | Bucket "bucket"
} representation keyed

type Child union {
  | Link "link"
  | HashMapNode "inline"
} representation keyed
```

So you end up with either:

```
{
  ...
  "data": [
    ... buckets and other child nodes here
    "child": {
      "link": CID
    },
    ... buckets and other child nodes here
}
```

or

```
{
  ...
  "data": [
    ... buckets and other child nodes here
    "child": {
      "node": {
        ... actual inline child node here
      }
    },
    ... buckets and other child nodes here
}
```

The other options we have are kinded (HashMapNode isn't a kind, it's a complex map representation), inline (won't work where one of these things is a kind, i.e. Link, it can only inline two complex map representations with a discriminator key), and envelope (would work but requires a discriminator key in addition to the child key). See docs from [here](https://github.com/ipld/specs/blob/master/schema-layer/schemas/representations.md#union-keyed-representation-example) down.

So we can't do a union of: `"child": CID` and `"child": { ... HashMapNode here ... }`. Which I know will bother at least @mikeal because it doesn't make block boundaries "transparent". But that's @warpfork's argument to have.

Reply:
@warpfork could this particular case could be solved by treating `HashMapNode` as a `map` for a kinded union?

```
type Child union {
  | Link link
  | HashMapNode map
} representation kinded
```

making

```
    "child": CID
```

or

```
    "child": { ... HashMapNode here }
```

is that approach a general solution to the "everything could be a link" desire?

Reply:
you beat me to the punch.  yes, we should be able to do exactly such an expression with a kinded union.

Reply:
OK, I've gone with that kinded form above, it's neat. I've also removed or altered explicit references to CIDs for read operations such that they could be inline. But I've also added this to the bottom of the summary in the top section of this doc:

> By default, each node in an IPLD HashMap is stored in a distinct IPLD block and CIDs are used for child node links. The schema and algorithm presented here also allows for inline child nodes rather than links, with read operations able to traverse multiple nodes within a single block where they are inlined. The production of inlined IPLD HashMaps is left unspecified and users should be aware that inlining breaks canonical form guarantees.

i.e. Inline is possible and reads should assume it's possible. How you write them and under what circumstances are up to you and they shouldn't be default mode. See 73ab06b for the precise changes I made on this point. Reviews appreciated.

Reply:
I'm perfectly happy with that phrasing.

Using a kinded union, as you've done, makes it super clear that we've got the possibility for "inline" vs separate linked blocks -- A+ good.

I also think -- if you want to -- you could get even more prescriptive than that text does:

Refining our exact language and understandings of this has been tricky.  My best thinking at present is that having "implicit links" in *general* was problematic (as it made canonicality more or less undefined)... but, we're not "in general", here; we're in the scope of an advanced layout.

Inside an advanced layout, we can set the definition of canonicality to whatever we want.  Because it's already scoped by the defn of the advanced layout, and since an advanced layout defn already includes descriptions of algorithms/decisions... it therefore follows that adding another decision description for whether or not to inline-vs-link isn't out of character at all; it fits right along with the rest.

In other words -- as long as the decision to inline-vs-link happens *inside* the scope of the advanced layout, rather than being unspecified and left to some higher level (aka, made into a problem for the user), we're good!